Upgrading to PostgreSQL 15 and moving to sclorg images

rooftopcellist opened the pull request:
##### SUMMARY

Upgrading from PostgreSQL 13 to 15.

kurokobo commented on config/manager/manager.yaml:
```diff
@@ -60,9 +60,9 @@ spec:
               value: quay.io/ansible/galaxy-ui:latest
             - name: RELATED_IMAGE_GALAXY_REDIS
-              value: redis:latest
+              value: quay.io/sclorg/redis-6-c9s:latest
```
Hi, perhaps this should be `quay.io/sclorg/redis-7-c9s:latest` instead of `6`.
I leave just one comment here but this is applicable for the changes for other files.

rooftopcellist replied:
@kurokobo for the time being, we need to stick with Redis 6 because of the matrix of where this code gets deployed. EL 8 does not provide Redis 7.  For simplicity and consistency, for now we have opted to stick with Redis 6 across the board, as there are no features we specifically need from Redis 7 and Redis 6 is stable.

kurokobo replied:
Ah make sense. Thanks for your clarifycation!

dsavineau replied:
not sure if that's a good idea to change the redis image in the same PR or at least it should be done in a separate commit because this will require more changes.
We currently don't provide any configmap for the redis configuration and the `redis:latest` image was using `/data` in its default configuration and that's what we're using for the redis PVC
https://github.com/docker-library/redis/blob/master/6.0/debian/Dockerfile#L144-L146
https://github.com/ansible/galaxy-operator/blob/main/roles/redis/templates/redis.deployment.yaml.j2#L55-L58

However, the redis image from sclorg uses `/var/lib/redis/data` as data directory
https://github.com/sclorg/redis-container/blob/master/6/root/usr/share/container-scripts/redis/common.sh#L8
https://github.com/sclorg/redis-container/blob/master/6/root/usr/share/container-scripts/redis/base.conf.template#L1

rooftopcellist replied:
@dsavineau good point.  I removed that from this PR.  I'll do that in a follow-up PR.

rooftopcellist commented:
I am about to be OOO for 2 weeks, so just a note to anyone who merges this.  We should not squash the commits here as some of the changes here we may want to port to other operators (specifically the change that checks both paths for PG_VERSION).

dsavineau commented on roles/postgres/defaults/main.yml:
```diff
@@ -1,5 +1,4 @@
 ---
-postgres_version: 13
 _postgres_image: postgres
```
Don't we have to change the postgresql container image value here too ?

rooftopcellist replied:
The `postgres_version` variable was replaced by `supported_pg_version` variable for consistency across the other operators.

dsavineau replied:
I'm talking about the other variable : `_postgres_image` ;)

rooftopcellist replied:
ahh, woops. Fixed now, thanks!

dsavineau commented on roles/postgres/tasks/upgrade_postgres.yml:
```diff
@@ -123,12 +122,34 @@
   set_fact:
     postgres_pod_name: "{{ postgres_pod['resources'][0]['metadata']['name'] }}"
 
+- name: Get the name of the service for the old postgres pod
+  k8s_info:
+    kind: Service
+    namespace: "{{ ansible_operator_meta.namespace }}"
+    label_selectors:
+      - "app.kubernetes.io/component=database"
+      - "app.kubernetes.io/instance=postgres-{{ ansible_operator_meta.name }}"
+      - "app.kubernetes.io/version={{ supported_pg_version }}"
+      - "app.kubernetes.io/managed-by={{ deployment_type }}-operator"
+  register: old_postgres_svc
+
+# TODO: The galaxy-operator labels are different. potential issue here.
+- name: Get the name of the service for the old postgres pod
+  k8s_info:
+    kind: Service
+    namespace: "{{ ansible_operator_meta.namespace }}"
+    label_selectors:
+      - "app.kubernetes.io/component=database"
+      - "app.kubernetes.io/instance={{ old_postgres_pod.metadata.labels['app.kubernetes.io/instance'] }}"
+      - "app.kubernetes.io/version={{ old_postgres_pod.metadata.labels['app.kubernetes.io/version'] }}"  # unique to the galaxy-operator
+      - "app.kubernetes.io/managed-by={{ deployment_type }}-operator"
+  register: old_postgres_svc
```
so we're trying to get information from the old service but we register to the same variable so anything registered in the initial task is erased.
Is that expected ? is this some preference order ?
Also the first task is using `"app.kubernetes.io/version={{ supported_pg_version }}"` but at this point we already have the `supported_pg_version` variable sets to 15 right ? in this case this isn't what we want, we want the old one from either postgresql 12 or 13 ?

rooftopcellist replied:
This first task should just be removed entirely, it was an intermediate task that never got removed. Good find.  I'll remove this now.

dsavineau commented on roles/redis/defaults/main.yml:
```diff
@@ -1,5 +1,5 @@
 ---
-_redis_image: redis:7
+_redis_image: redis:latest
```
You should revert this change to 7 to avoid bumping redis to a major release if something is release in the future

rooftopcellist replied:
Agreed, I just updated the PR. Thanks!

aknochow commented:
Looking into why those 2 checks are failing...

dsavineau commented:
@rooftopcellist I rebased this PR and added two commits (same as eda operator) in order to get the upgrade working

kurokobo commented:
Hi, thanks for working on this!

In changing PSQL from docker.io to sclorg, some considerations need to be made.

- The path where PV should be mounted is `/var/lib/pgsql/data`, not `/var/lib/postgresql/data`
  - https://github.com/ansible/awx-operator/pull/1766
  - This causes data loss, and this is why AWX Operator 2.13.0 is marked as DO NOT USE and removed from quay.io
- Data path cannot be changed using the `PGDATA` environment variable
  - https://github.com/ansible/awx-operator/issues/1790
- `postgres_data_path` should be fixed to `/var/lib/pgsql/data/userdata` and should be ignored if `postgres_data_path` is modified in Galaxy CR. If the PV is mounted on a path that is not actually used, data persistence may be broken.
  - `postgres_data_path` should be deprecated in the CRD, and docs should be updated
  - As a test, restart the PSQL pod after installation to make sure no data is lost.
- The executing user is UID 26, not UID 0. This may cause permission issues for subpath data
  - https://github.com/ansible/awx-operator/issues/1770

AWX Operator 2.13.0 has adopted sqlorg's PSQL, but there are a number of open issues, so it may be safer for Galaxy Operator and EDA Server Operator to collectively address them after things settle down.
EDA Server Operator has been using sclorg's PSQL 13 from the start, so the `postgres_data_path` problem seems unlikely.

rooftopcellist commented:
Thank you for the patches @dsavineau !

@kurokobo Thank for your notes.  I think I have addressed all of these now:
* The default postgres data path is now `/var/lib/pgsql/data/userdata`
* Since the `PGDATA` environment variable is no longer respected by the sclorg image, I removed the ability to configure it as we did in the awx-operator.  If a user has `postgres_data_path` defined on their Galaxy spec, it will be ignored (the k8s validator omits it silently when Galaxy CR is applied).

> Note: I confirmed that in the sclorg and registry.redhat.io provided PostgreSQL images, the data dir path that is actually used is `/var/lib/pgsql/data/userdata/`, even though PGDATA gets set to the value passed into the StatefulSet yaml.  So we no longer specify the PGDATA env var to avoid confusion, since it is not respected by the PostgreSQL image anyways. 

I still need to try to reproduce the UID 26 permissions issue you mentioned ([link](https://github.com/ansible/awx-operator/issues/1770)), I don't see it on my openshift cluster, but that makes sense  because permissions are handled differently there.  I'll try to reproduce on a k8s cluster.

kurokobo commented:
Just added link to my comment on the issue: https://github.com/ansible/awx-operator/issues/1770#issuecomment-2031118430

rooftopcellist commented:
I have now added in the `postgres_data_path` related changes from this PR:
* https://github.com/ansible/awx-operator/pull/1798

I also just included the changes to make it possible initialize the postgresql data volume with the correct permissions if running on k8s with a hostMounted volume.  
* https://github.com/ansible/awx-operator/pull/1805/

kurokobo commented on docs/user-guide/database-configuration.md:
````diff
@@ -0,0 +1,114 @@
+type: Opaque
+```
+
+> Please ensure that the value for the variable `password` should _not_ contain single or double quotes (`'`, `"`) or backslashes (`\`) to avoid any issues during deployment, [backup](./roles/backup) or [restoration](./roles/restore).
````
```suggestion
> Please ensure that the value for the variable `password` should _not_ contain single or double quotes (`'`, `"`) or backslashes (`\`) to avoid any issues during deployment, [backup](https://github.com/ansible/galaxy-operator/tree/main/roles/backup) or [restoration](https://github.com/ansible/galaxy-operator/tree/main/roles/restore).
```

rooftopcellist replied:
Thanks! I just added the full link for those.

rooftopcellist commented:
@dsavineau @kurokobo @aknochow This PR is now ready for merge IMO.  I was able to do a fresh install, upgrade (main --> pg15 branch) and backup and restore on k3s using mounted volumes.  There are details in the commit history for the changes needed and why.

dsavineau commented on roles/backup/tasks/remove_management_pod.yml:
```diff
@@ -0,0 +1,8 @@
+- name: Delete any existing management pod
+  k8s:
+    name: "{{ ansible_operator_meta.name }}-backup-manager"
+    kind: Pod
```
Maybe not to handle in this PR but as a follow
We should always define the `api_version` parameter with the k8s to avoid some issue with the default api_version value (see some awx-operator about this)

dsavineau commented on roles/backup/templates/backup-content-k8s-job.yaml.j2:
```diff
@@ -0,0 +1,50 @@
+{% endif %}
+      containers:
+      - name: backup-content
+        image: quay.io/ansible/galaxy-ng:latest # TODO: Add set image tasks, and var ref here, etc.
```
Do we want to switch this value to the correct variable name in this PR ?
Also do we need to handle resource limit on this container ?

rooftopcellist replied:
Agreed, thanks for calling these points out!  I just addressed both in the commit I just pushed.

dsavineau commented on config/rbac/role.yaml:
```diff
@@ -91,6 +91,19 @@ rules:
+  - apiGroups:
+      - batch
+    resources:
+      - cronjobs
```
We don't need the cronjobs resource only the jobs one

```suggestion
```

dsavineau commented on roles/galaxy-content/templates/galaxy-content.deployment.yaml.j2:
```diff
@@ -41,15 +41,28 @@ spec:
         app.kubernetes.io/managed-by: '{{ deployment_type }}-operator'
+      annotations:
+        kubectl.kubernetes.io/default-container: content
+{% for template in [
+    "galaxy-file-storage.pvc",
+  ] %}
+        checksum-{{ template | replace('/', '-') }}: "{{ lookup('template', template + '.yaml.j2') | sha1 }}"
+{% endfor %}
```
we should only do this template if the storage backend if file

rooftopcellist commented:
Ok, testing went well.  There is one caveat, upstream users on k3s will have a manual step:

If you see the following error in the postgres pod's logs upon upgrade, or when running a restore after the fact, then  you are likely running on k3s and need to enable the init container to set permissions on the postgresql directory. 

```
$ kubectl -n galaxy logs statefulset/new-galaxy-postgres-15
mkdir: cannot create directory '/var/lib/pgsql/data/userdata': Permission denied
```

 Follow these steps to remediate the issue by setting the postgres_data_volume_init parameter true and deleting the new postgresql StatefulSet.

```
# Patch your Galaxy custom resource
kubectl -n galaxy patch galaxy <deployment-name> --type=merge -p '{"spec": {"postgres_data_volume_init": true}}'

# Delete the new postgres stateful set
kubectl -n galaxy delete statefulset <deployment-name>-postgres-15
```

cc @kurokobo 

I will be sure to include a note in the release notes. 

As we are ready to merge this, I will cut an upstream release first since a few things have merged since the last one, so that community members could get those fixes without doing the database upgrade if desired.

kurokobo commented:
@rooftopcellist 
Thanks for pointing that out. Currently my guide does not include any upgrade guide for Galaxy Operator (I have only for AWX Operator), but you are right, some manualy interventions are required.
For AWX Operator, I was provided the guide to create new PV for PSQL 15 that includes `chown 26:0` command, so if the users read my guide carefully (I believe so 😃 ), I believe users never face the issue: https://github.com/kurokobo/awx-on-k3s/blob/main/tips/upgrade-operator.md

Anyway I will give new Galaxy Operator try in next days. Thanks!

rooftopcellist commented:
Ahh, even better!  And should they miss a step and end up needing the initContainer to set the permissions to get the deployment back in a good state, they have the remediation steps on my earlier comment, which I'll include in the release notes.  

Thanks for taking a look!

rooftopcellist commented:
I plan to cut another release including the changes in this PR tomorrow morning.

kurokobo commented:
@rooftopcellist 
Made a minimal test and seems both 2024.4.30 and 2024.5.1 work fine😃 I've updated my guide to support 2024.5.1.
Thanks for working hard and releasing this with detailed release notes!

rooftopcellist commented on roles/restore/tasks/deploy_galaxy.yml:
```diff
@@ -14,66 +14,27 @@
   set_fact:
-    cr_spec: "{{ cr_object.stdout }}"
+    cr_spec_from_backup: "{{ cr_object.stdout }}"
     cr_spec_strip: "{ "
```
We should remove this `cr_spec_strip` var, it looks like it is no longer used.

rooftopcellist replied:
Same with the other variables here that seem to have been for building the cr_spec... unless I've misunderstood this.
